test: Add unit tests for package/utils

hampuslavin opened the pull request:
Adds tests as outlined in #3308. 

The file is now at ~95% test coverage. However, it's always a question with regard to the quality of that percentage. Do you feel these tests increase the confidence that the exported utils are indeed working as intended?

I could not utilise the test helper `API.createElement` since it caused the snapshots to vary between tests, hence I did a small local factory helper. Maybe there's another way.

Additionally - I have some questions about the type `ExportOpts`:
 - Any particular reason default values are provided for `getDimensions` although it's not an optional prop?
 - How would you like to handle the fact that `getDimensions` is not used (not needed?) by `exportToSvg`? In this first draft I just updated the return type signature.
 
 Cheers!

ad1992 commented on src/tests/packages/utils.test.ts:
```diff
@@ -0,0 +1,127 @@
+    expect(canvas.height).toBe(200);
+  });
+});
+describe("exportToBlob", () => {
```
```suggestion

describe("exportToBlob", () => {
```

hampuslavin replied:
✅

ad1992 commented on src/tests/packages/utils.test.ts:
```diff
@@ -0,0 +1,127 @@
+
+describe("exportToCanvas", () => {
+  const EXPORT_PADDING = 10;
+  test("with default arguments", () => {
```
```suggestion
  it("with default arguments", () => {
```

hampuslavin replied:
✅

ad1992 commented on src/tests/packages/utils.test.ts:
```diff
@@ -0,0 +1,127 @@
+    expect(canvas.width).toBe(100 + 2 * EXPORT_PADDING);
+    expect(canvas.height).toBe(100 + 2 * EXPORT_PADDING);
+  });
+  test("when custom width and height", () => {
```
```suggestion
  it("when custom width and height", () => {
```

for consistency

hampuslavin replied:
Absolutely ✅

ad1992 commented on src/tests/packages/utils.test.ts:
```diff
@@ -0,0 +1,127 @@
+      });
+      expect(blob?.type).toBe("image/jpeg");
+    });
+    it("should default to image/png", async () => {
```
```suggestion
    
    it("should default to image/png", async () => {
```
lets add new line between individual specs for better readability

hampuslavin replied:
✅

ad1992 commented on src/tests/packages/utils.test.ts:
```diff
@@ -0,0 +1,127 @@
+        elementOverrides: { width: 100, height: 100 },
+      }),
+    });
+    expect(svgElement).toHaveAttribute("height", "120");
+    expect(svgElement).toHaveAttribute("width", "120");
+    expect(svgElement).toHaveAttribute("viewBox", "0 0 120 120");
```
not needed since it will be covered in snapshot

hampuslavin replied:
✅

ad1992 commented on src/packages/utils.ts:
```diff
@@ -83,7 +83,7 @@ export const exportToSvg = ({
   exportPadding,
   metadata,
-}: ExportOpts & {
+}: Omit<ExportOpts, "getDimensions"> & {
```
Agreed we should omit `getDimensions` since its not getting used

hampuslavin replied:
Great!

ad1992 reviewed (changes requested):
Thanks a lot for adding the specs @hampuslavin ❤️ 
> The file is now at ~95% test coverage. However, it's always a question with regard to the quality of that percentage. Do you feel these tests increase the confidence that the exported utils are indeed working as intended

Yep these specs will give us confidence for sure as in consumers it will be used as utils so making any change in future, the unit tests will definitely help us.
However some utils for example `exportToSvg` is just a wrapper over the original `exportToSvg` and specs also are testing the original util so we might want to move some specs to `util.test.ts` and only check if the original export function is called with correct args in package utils `exportToSvg`.
Also let's cover the watermark and dark mode cases as well.

> I could not utilise the test helper API.createElement since it caused the snapshots to vary between tests, hence I did a small local factory helper. Maybe there's another way.

Yeah we don't need to use the API here, fixtures are fine as we are unit testing
>Any particular reason default values are provided for getDimensions although it's not an optional prop?

This should be an optional prop indeed.

> How would you like to handle the fact that getDimensions is not used (not needed?) by exportToSvg? In this first draft I just updated the return type signature.

Yep what you did is fine since its not getting used 👍

hampuslavin commented on src/tests/packages/utils.test.ts:
```diff
@@ -0,0 +1,127 @@
+  boundElementIds: null,
+  ...overrides,
+});
+const diagramFactory = ({ overrides = {}, elementOverrides = {} } = {}) => ({
```
✅

hampuslavin commented on src/tests/packages/utils.test.ts:
```diff
@@ -0,0 +1,127 @@
+import { ExcalidrawElement } from "../../element/types";
+import * as utils from "../../packages/utils";
+
+const elementFactory = (overrides = {}): ExcalidrawElement => ({
```
Moved to fixtures and created some basic types, it's easy to add more if need be ✅

hampuslavin commented:
@ad1992  Thank you for the review! 🌟 
- Moved the `exportToSvg` tests to `tests/util.test.ts` and added cases accordingly. 
- Put fixtures in separate files
- Fixed formatting

I'll add the last unit tests later this week :) 👍

hampuslavin commented:
Alright, now I've added tests for package/utils exportToSvg that are testing the wrapper rather than the integration. 

Happy to take any comments and iterate again if needed! 🙂 

Cheers

ad1992 commented on src/tests/packages/utils.test.ts:
```diff
@@ -0,0 +1,121 @@
+    };
+
+    expect(passedElements().length).toBe(3);
+    expect(passedOptionsWhenDefault).toMatchSnapshot();
```
do we need to check snapshot here ? snapshot would make more sense in actual util, here we can just check if the function was called with correct args instead.

ad1992 replied:
ok I see the default app state is what makes it so big, ok let it be then, probably we should fix the util itself so entire default app state is not spread but thats for another pr

hampuslavin replied:
Ok! 👍

ad1992 commented on src/tests/utils.test.ts:
```diff
@@ -1,4 +1,7 @@
+import { NonDeletedExcalidrawElement } from "../element/types";
+import * as exportUtils from "../scene/export";
```
lets rename this file to `export.test.ts` instead to match the original file name.

hampuslavin replied:
Alright! 👍

ad1992 commented on src/tests/utils.test.ts:
```diff
@@ -11,3 +14,101 @@ describe("Test isTransparent", () => {
+
+  it("with dark mode", () => {
+    const svgElement = exportUtils.exportToSvg(ELEMENTS, {
+      exportBackground: false,
+      viewBackgroundColor: "#ffffff",
+      exportWithDarkMode: true,
+      shouldAddWatermark: false,
```
can we have some default args above so we can just spread over it instead of passing everytime ?

ad1992 commented:
@hampuslavin thanks a lot for adding the specs ❤️.
